File: apps/mobile/src/features/voice/live/pcm.ts
/**
 * Audio format glue between react-native-audio-api (float32 samples, −1..1) and
 * the Gemini Live API (16-bit PCM, little-endian: 16 kHz mono in, 24 kHz out).
 * All pure + dependency-free so it can be unit-tested off-device.
 */

const B64 = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/';
const B64_LOOKUP = (() => {
  const table = new Uint8Array(256);
  for (let i = 0; i < B64.length; i++) table[B64.charCodeAt(i)] = i;
  return table;
})();

/** Float32 [-1,1] → little-endian int16 PCM bytes. */
export function floatTo16BitPCM(input: Float32Array): Uint8Array {
  const out = new Uint8Array(input.length * 2);
  const view = new DataView(out.buffer);
  for (let i = 0; i < input.length; i++) {
    const s = Math.max(-1, Math.min(1, input[i]));
    view.setInt16(i * 2, s < 0 ? s * 0x8000 : s * 0x7fff, true /* little-endian */);
  }
  return out;
}

/** Little-endian int16 PCM bytes → Float32 [-1,1]. */
export function pcm16ToFloat32(bytes: Uint8Array): Float32Array {
  const frames = Math.floor(bytes.length / 2);
  const out = new Float32Array(frames);
  const view = new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength);
  for (let i = 0; i < frames; i++) {
    out[i] = view.getInt16(i * 2, true) / 0x8000;
  }
  return out;
}

/**
 * Linear-interpolation resample. Only needed if the recorder won't honor the
 * requested 16 kHz (the docs warn the hardware rate may differ). Fine for a
 * spike; swap for a filtered resampler if aliasing is audible.
 */
export function resampleFloat32(input: Float32Array, fromRate: number, toRate: number): Float32Array {
  if (fromRate === toRate) return input;
  const ratio = fromRate / toRate;
  const outLength = Math.floor(input.length / ratio);
  const out = new Float32Array(outLength);
  for (let i = 0; i < outLength; i++) {
    const pos = i * ratio;
    const i0 = Math.floor(pos);
    const i1 = Math.min(i0 + 1, input.length - 1);
    const frac = pos - i0;
    out[i] = input[i0] * (1 - frac) + input[i1] * frac;
  }
  return out;
}

export function bytesToBase64(bytes: Uint8Array): string {
  let out = '';
  let i = 0;
  for (; i + 2 < bytes.length; i += 3) {
    const n = (bytes[i] << 16) | (bytes[i + 1] << 8) | bytes[i + 2];
    out += B64[(n >> 18) & 63] + B64[(n >> 12) & 63] + B64[(n >> 6) & 63] + B64[n & 63];
  }
  const rem = bytes.length - i;
  if (rem === 1) {
    const n = bytes[i] << 16;
    out += B64[(n >> 18) & 63] + B64[(n >> 12) & 63] + '==';
  } else if (rem === 2) {
    const n = (bytes[i] << 16) | (bytes[i + 1] << 8);
    out += B64[(n >> 18) & 63] + B64[(n >> 12) & 63] + B64[(n >> 6) & 63] + '=';
  }
  return out;
}

export function base64ToBytes(b64: string): Uint8Array {
  const clean = b64.replace(/[^A-Za-z0-9+/]/g, '');
  const pad = clean.endsWith('==') ? 2 : clean.endsWith('=') ? 1 : 0;
  const byteLength = (clean.length * 3) / 4 - pad;
  const out = new Uint8Array(byteLength);
  let p = 0;
  for (let i = 0; i < clean.length; i += 4) {
    const n =
      (B64_LOOKUP[clean.charCodeAt(i)] << 18) |
      (B64_LOOKUP[clean.charCodeAt(i + 1)] << 12) |
      (B64_LOOKUP[clean.charCodeAt(i + 2)] << 6) |
      B64_LOOKUP[clean.charCodeAt(i + 3)];
    if (p < byteLength) out[p++] = (n >> 16) & 0xff;
    if (p < byteLength) out[p++] = (n >> 8) & 0xff;
    if (p < byteLength) out[p++] = n & 0xff;
  }
  return out;
}

File: apps/ios/ReciPeer/Features/Voice/PCM.swift
import Foundation

/// Audio format glue between AVAudioEngine (float32 samples, −1..1) and the
/// Gemini Live API (16-bit PCM, little-endian: 16 kHz mono in, 24 kHz out).
/// Mirrors live/pcm.ts.

/// Float32 [-1,1] → little-endian int16 PCM bytes.
func floatTo16BitPCM(_ input: [Float]) -> [UInt8] {
    var out = [UInt8]()
    out.reserveCapacity(input.count * 2)
    for sample in input {
        let clamped = max(-1, min(1, sample))
        let value = Int16(clamped < 0 ? clamped * Float(Int16.max) : clamped * Float(Int16.max))
        withUnsafeBytes(of: value.littleEndian) { out.append(contentsOf: $0) }
    }
    return out
}

/// Little-endian int16 PCM bytes → Float32 [-1,1].
func pcm16ToFloat32(_ bytes: [UInt8]) -> [Float] {
    let frames = bytes.count / 2
    var out = [Float]()
    out.reserveCapacity(frames)
    var index = 0
    for _ in 0..<frames {
        let value = Int16(bytes[index]) | (Int16(bytes[index + 1]) << 8)
        out.append(Float(value) / Float(Int16.max))
        index += 2
    }
    return out
}

/// Linear-interpolation resample (pcm.ts resampleFloat32).
func resampleFloat32(_ input: [Float], fromRate: Double, toRate: Double) -> [Float] {
    guard fromRate != toRate, !input.isEmpty else { return input }
    let ratio = fromRate / toRate
    let outLength = Int(Double(input.count) / ratio)
    var out = [Float](repeating: 0, count: outLength)
    for i in 0..<outLength {
        let pos = Double(i) * ratio
        let i0 = Int(pos)
        let i1 = min(i0 + 1, input.count - 1)
        let frac = Float(pos - Double(i0))
        out[i] = input[i0] * (1 - frac) + input[i1] * frac
    }
    return out
}

func bytesToBase64(_ bytes: [UInt8]) -> String {
    Data(bytes).base64EncodedString()
}

func base64ToBytes(_ b64: String) -> [UInt8] {
    Data(base64Encoded: b64).map { Array($0) } ?? []
}

/// Parse "audio/pcm;rate=24000" → 24000 (live-playback.ts rateFromMime).
func rateFromMime(_ mime: String) -> Double {
    if let range = mime.range(of: #"rate=(\d+)"#, options: .regularExpression) {
        let digits = mime[range].dropFirst("rate=".count)
        return Double(digits) ?? 24000
    }
    return 24000
}
